File: src/page/info/filter/UnitListTable.java
package page.info.filter;

import common.battle.Basis;
import common.battle.BasisSet;
import common.battle.data.MaskUnit;
import common.system.Node;
import common.util.unit.Enemy;
import common.util.unit.Form;
import common.util.unit.Unit;
import page.MainFrame;
import page.Page;
import page.info.UnitInfoPage;
import page.support.SortTable;
import page.support.UnitTCR;

import java.awt.*;

public class UnitListTable extends SortTable<Form> {

    private static final long serialVersionUID = 1L;

    private static String[] tit;

    static {
        redefine();
    }

    public static void redefine() {
        tit = new String[]{"ID", "name", Page.get(1, "pref"), "HP", "HB", "atk", Page.get(1, "range"),
                Page.get(1, "speed"), "dps", Page.get(1, "preaa"), "CD", Page.get(1, "price"), Page.get(1, "atkf")};
    }

    private final Page page;

    public UnitListTable(Page p) {
        page = p;

        setDefaultRenderer(Enemy.class, new UnitTCR(lnk));

    }

    public void clicked(Point p) {
        if (list == null)
            return;
        int c = getColumnModel().getColumnIndexAtX(p.x);
        c = lnk[c];
        int r = p.y / getRowHeight();
        if (r < 0 || r >= list.size() || c != 1)
            return;
        Form e = list.get(r);
        Node<Unit> n = Node.getList(list, e);
        MainFrame.changePanel(new UnitInfoPage(page, n));

    }

    @Override
    public Class<?> getColumnClass(int c) {
        c = lnk[c];
        if (c == 1)
            return Enemy.class;
        return String.class;
    }

    @Override
    protected int compare(Form e0, Form e1, int c) {
        if (c == 0) {
            int val = e0.uid.compareTo(e1.uid);
            return val != 0 ? val : Integer.compare(e0.fid, e1.fid);
        }
        if (c == 1)
            return e0.toString().compareTo(e1.toString());
        int i0 = (int) get(e0, c);
        int i1 = (int) get(e1, c);
        return i0 > i1 ? 1 : i0 == i1 ? 0 : -1;
    }

    @Override
    protected Object get(Form e, int c) {
        Basis b = BasisSet.current();
        MaskUnit du = e.maxu();
        double mul = e.unit.lv.getMult(e.unit.getPrefLv());
        double atk = b.t().getAtkMulti();
        double def = b.t().getDefMulti();
        int itv = du.getItv();
        if (c == 0)
            return e.uid + "-" + e.fid;
        else if (c == 1)
            return e;
        else if (c == 2)
            return e.unit.getPrefLv();
        else if (c == 3)
            return (int) (du.getHp() * mul * def);
        else if (c == 4)
            return du.getHb();
        else if (c == 5)
            return (int) (du.allAtk() * mul * atk);
        else if (c == 6)
            return du.getRange();
        else if (c == 7)
            return du.getSpeed();
        else if (c == 8)
            return (int) (du.allAtk() * mul * atk * 30 / itv);
        else if (c == 9)
            return du.rawAtkData()[0][1];
        else if (c == 10)
            return b.t().getFinRes(du.getRespawn());
        else if (c == 11)
            return e.getDefaultPrice(1);
        else if (c == 12)
            return du.getItv();
        else
            return null;
    }

    @Override
    protected String[] getTit() {
        return tit;
    }

}

File: src/page/info/filter/UnitListTable.kt
package page.info.filter

import common.battle.Basis
import common.battle.BasisSet
import common.system.Node
import common.util.unit.Enemy
import common.util.unit.Form
import common.util.unit.Unit
import page.MainFrame
import page.Page
import page.info.UnitInfoPage
import page.support.SortTable
import page.support.UnitTCR
import java.awt.Point

class UnitListTable(private val page: Page) : SortTable<Form?>() {
    companion object {
        private const val serialVersionUID = 1L
        private var tit: Array<String>
        fun redefine() {
            tit = arrayOf("ID", "name", Page.Companion.get(1, "pref"), "HP", "HB", "atk", Page.Companion.get(1, "range"),
                    Page.Companion.get(1, "speed"), "dps", Page.Companion.get(1, "preaa"), "CD", Page.Companion.get(1, "price"), Page.Companion.get(1, "atkf"))
        }

        init {
            redefine()
        }
    }

    fun clicked(p: Point) {
        if (list == null) return
        var c: Int = getColumnModel().getColumnIndexAtX(p.x)
        c = lnk.get(c)
        val r: Int = p.y / getRowHeight()
        if (r < 0 || r >= list.size || c != 1) return
        val e: Form = list.get(r)
        val n: Node<Unit> = Node.Companion.getList(list, e)
        MainFrame.Companion.changePanel(UnitInfoPage(page, n))
    }

    override fun getColumnClass(c: Int): Class<*> {
        var c = c
        c = lnk.get(c)
        return if (c == 1) Enemy::class.java else String::class.java
    }

    protected override fun compare(e0: Form, e1: Form, c: Int): Int {
        if (c == 0) {
            val `val` = e0.uid.compareTo(e1.uid)
            return if (`val` != 0) `val` else Integer.compare(e0.fid, e1.fid)
        }
        if (c == 1) return e0.toString().compareTo(e1.toString())
        val i0 = get(e0, c) as Int
        val i1 = get(e1, c) as Int
        return if (i0 > i1) 1 else if (i0 == i1) 0 else -1
    }

    protected override operator fun get(e: Form, c: Int): Any? {
        val b: Basis = BasisSet.Companion.current()
        val du: MaskUnit = e.maxu()
        val mul = e.unit.lv.getMult(e.unit.prefLv)
        val atk: Double = b.t().getAtkMulti()
        val def: Double = b.t().getDefMulti()
        val itv: Int = du.getItv()
        return if (c == 0) e.uid.toString() + "-" + e.fid else if (c == 1) e else if (c == 2) e.unit.prefLv else if (c == 3) (du.getHp() * mul * def) else if (c == 4) du.getHb() else if (c == 5) (du.allAtk() * mul * atk) else if (c == 6) du.getRange() else if (c == 7) du.getSpeed() else if (c == 8) (du.allAtk() * mul * atk * 30 / itv) else if (c == 9) du.rawAtkData().get(0).get(1) else if (c == 10) b.t().getFinRes(du.getRespawn()) else if (c == 11) e.getDefaultPrice(1) else if (c == 12) du.getItv() else null
    }

    override fun getTit(): Array<String> {
        return tit
    }

    init {
        setDefaultRenderer(Enemy::class.java, UnitTCR(lnk))
    }
}
